### PR TITLE
#1008: Added Q.WATCH support to Websocket

### LINKED ISSUE #1008: Add QWatch support to Websocket
This is a follow up work on PR #690

QWatch is currently supported on cli and http

### PR DESCRIPTION
**Pull request by psrvere**
Closes #1008 

This PR
1. adds support for Q.WATCH command on Websocket server
2. changes Websocket server logic
    a. updates `WebsocketServer` struct to use `ioChan` and `qwatchResponseChan`
    b. extracts out response processing logic to a function
    c. adds a function to process Q.WATCH updates
3. changes command parsing logic in `ParseWebsocketMessage` to support Q.WATCH command
4. adds more unit tests to increase coverage of parsing logic
5. adds unit tests for Q.WATCH
6. adds integration tests for Q.WATCH
7. updates Q.WATCH docs

### CODE REVIEW

**Comment by psrvere** [comment]
@lucifercr07 @pratikpandey21 - Added Qwatch to Websocket. Please review.

File: internal/server/utils/redisCmdAdapter_test.go [hunk 1]
```diff
@@ -284,6 +386,18 @@ func TestParseWebsocketMessage(t *testing.T) {
 			expectedCmd:  "JSON.INGEST",
 			expectedArgs: []string{"", "$..field", `{"field":"value"}`},
 		},
+		{
+			name:         "Test simple Q.WATCH command",
+			message:      "q.watch \"select $key, $value where $key like 'k?'\"",
+			expectedCmd:  "Q.WATCH",
```
Comment (pratikpandey21) [inline internal/server/utils/redisCmdAdapter_test.go:392]: Shouldn't this be QWATCH? Why Q.WATCH?

**Review by pratikpandey21 — COMMENTED** [review]
I'd recommend waiting for https://github.com/DiceDB/dice/pull/1086
and leverage the code there as that would enable multithreading in WS.

Will review once those changes are there

File: internal/server/utils/redisCmdAdapter_test.go (hunk 1 above)
Comment (JyotinderSingh) [reply internal/server/utils/redisCmdAdapter_test.go:392]: We have renamed the command to be consistent with GET.WATCH, ZRANGE.WATCH, etc.

Comment (psrvere) [reply internal/server/utils/redisCmdAdapter_test.go:392]: QWATCH has been renamed to Q.WATCH everywhere in the codebase in this PT #1102

File: internal/server/websocketServer.go [hunk 2]
```diff
@@ -130,20 +123,22 @@ func (s *WebsocketServer) WebsocketHandler(w http.ResponseWriter, r *http.Reques
...
 		if err != nil {
-			writeResponse(conn, []byte("error: command reading failed"))
-			continue
+			break
 		}
 
 		// parse message to dice command
 		diceDBCmd, err := utils.ParseWebsocketMessage(msg)
 		if errors.Is(err, diceerrors.ErrEmptyCommand) {
 			continue
```
[the remark this answers is not in the document]
Comment (psrvere) [reply internal/server/websocketServer.go:137]: Added a warning here.

File: internal/server/websocketServer.go [hunk 3]
```diff
@@ -153,78 +148,140 @@ func (s *WebsocketServer) WebsocketHandler(w http.ResponseWriter, r *http.Reques
 		}
```
[the remark this answers is not in the document]
Comment (psrvere) [reply internal/server/websocketServer.go:149]: Sure. Added a comment for now to capture this point.

File: internal/server/websocketServer.go
[the remark this answers is not in the document]
Comment (psrvere) [reply internal/server/websocketServer.go:1]: Done

File: internal/server/utils/redisCmdAdapter_test.go (hunk 1 above)
Comment (psrvere) [reply internal/server/utils/redisCmdAdapter_test.go:392]: Done.

File: internal/server/websocketServer.go [hunk 4]
```diff
@@ -153,78 +148,140 @@ func (s *WebsocketServer) WebsocketHandler(w http.ResponseWriter, r *http.Reques
...
-			"(nil)",  // Represents a RESP Nil Bulk String, which indicates a null value.
-			"OK",     // Represents a RESP Simple String with value "OK".
-			"QUEUED", // Represents a Simple String indicating that a command has been queued.
-			"0",      // Represents a RESP Integer with value 0.
-			"1",      // Represents a RESP Integer with value 1.
-			"-1",     // Represents a RESP Integer with value -1.
-			"-2",     // Represents a RESP Integer with value -2.
-			"*0",     // Represents an empty RESP Array.
+			// start a goroutine for subsequent updates
+			go s.processQwatchUpdates(clientIdentifierID, conn, diceDBCmd)
```
[the remark this answers is not in the document]
Comment (psrvere) [reply internal/server/websocketServer.go:177]: Yes, this is absolutely critical. I realised this while writing tests with multiple subscribers. Have created a follow up issue #1153

File: integration_tests/commands/websocket/qwatch_test.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
+package websocket
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestQWatch(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (psrvere) [reply integration_tests/commands/websocket/qwatch_test.go:9]: Added the test but with single subscriber right now. As part of #1153, I will increase the number of subscriber to 5 (arbitrary)

[the remark this answers is not in the document]
Comment (psrvere) [reply integration_tests/commands/websocket/qwatch_test.go:9]: This test also will come as part of #1153. Currently previous tests leave a subscriber which also receive response when value of a key is updated. So it become multi subscriber problem again.

File: internal/server/websocketServer.go [hunk 6]
```diff
@@ -153,78 +148,140 @@ func (s *WebsocketServer) WebsocketHandler(w http.ResponseWriter, r *http.Reques
...
+		select {
+		case resp := <-s.qwatchResponseChan:
+			if resp.ClientIdentifierID == clientIdentifierID {
+				if err := s.processResponse(conn, dicDBCmd, resp); err != nil {
+					s.logger.Debug("Error writing response to client. Shutting down goroutine for qwatch updates", slog.Any("clientIdentifierID", clientIdentifierID), slog.Any("error", err))
+					return
+				}
 			}
+		case <-s.shutdownChan:
+			return
```
[the remark this answers is not in the document]
Comment (psrvere) [reply internal/server/websocketServer.go:199]: Sure. Let me also figure out how q.unwatch will work in websocket. Will address this comment as part of #1154

**Comment by psrvere** [comment]
@lucifercr07 - have addressed all comments and created new issues where needed. Please review again.

File: internal/server/websocketServer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucifercr07) [reply internal/server/websocketServer.go:142]: @psrvere not able to see the log statement added here, is commit pushed?

**Review by lucifercr07 — CHANGES_REQUESTED** [review]

File: internal/server/websocketServer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (psrvere) [reply internal/server/websocketServer.go:142]: @lucifercr07 - the warning was added as part of #1142 - please see [here](https://github.com/DiceDB/dice/pull/1142/files#diff-2f53c4849c012559c5361197c1ce70c4c56e8a9a5c8e0e46f41d8f3c61ee9bf2)

That was important for integration tests to work so had to move to that PR.

**Comment by psrvere** [comment]
@lucifercr07 - merging with latest master and resolve a few conflicts.

**Comment by psrvere** [comment]
@lucifercr07 - good to go, please check.

**Review by lucifercr07 — APPROVED** [review]
LGTM

**Pull request merged**